frontend/packages/console-shared/src/hooks/useUserSettings.ts

```diff
@@ -0,0 +1,112 @@
+const USER_SETTING_CONFIGMAP_NAMESPACE = 'console-user-settings';
+
+export const useUserSettings = (key: string, defaultValue: string = '') => {
+  const userUid = useSelector(
+    (state: RootState) => state.UI.get('user')?.metadata?.uid ?? 'kubeadmin',
+  );
+
+  const [configMapResource, configMapData] = React.useMemo(
+    () => [
+      {
+        kind: ConfigMapModel.kind,
+        namespace: USER_SETTING_CONFIGMAP_NAMESPACE,
+        isList: false,
+        name: `user-settings-${userUid}`,
+      },
+      {
+        apiVersion: ConfigMapModel.apiVersion,
+        kind: ConfigMapModel.kind,
+        metadata: {
+          name: `user-settings-${userUid}`,
+          namespace: USER_SETTING_CONFIGMAP_NAMESPACE,
+        },
+        data: {
+          [key]: defaultValue,
+        },
+      },
+    ],
+    [userUid, key, defaultValue],
+  );
+
+  const [cfData, cfLoaded, cfLoadError] = useK8sWatchResource<K8sResourceKind>(configMapResource);
+  const [settings, setSettings] = React.useState('');
+
+  // This won't be needed once we have proxy api
+  const getProject = React.useCallback(async () => {
+    try {
+      await k8sGet(ProjectModel, USER_SETTING_CONFIGMAP_NAMESPACE);
+    } catch {
+      await k8sCreate(ProjectRequestModel, {
+        metadata: {
+          name: USER_SETTING_CONFIGMAP_NAMESPACE,
+        },
+      });
+    }
+  }, []);
+
+  const createConfigMap = React.useCallback(async () => {
+    try {
+      await k8sCreate(ConfigMapModel, configMapData);
+    } catch (err) {
+      // eslint-disable-next-line no-console
+      console.error(err);
+    }
+  }, [configMapData]);
+
+  const patchConfigMap = React.useCallback(
+    async (patch) => {
+      try {
+        await k8sPatch(ConfigMapModel, cfData, patch);
+      } catch (err) {
+        // eslint-disable-next-line no-console
+        console.error(err);
+      }
+    },
+    [cfData],
+  );
+
+  React.useEffect(() => {
+    if (cfLoadError || (!cfData && cfLoaded) || (cfLoadError === undefined && !cfLoaded)) {
+      (async () => {
+        await getProject();
+        await createConfigMap();
+      })();
+    } else if (cfData && cfLoaded && settings !== cfData.data?.[key]) {
+      setSettings(cfData.data?.[key] ?? defaultValue);
+    }
+    // This effect should only be run once per key.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [key, cfLoaded]);
+
+  React.useEffect(() => {
+    if (cfData && cfLoaded && settings) {
+      const patch = [
+        {
+          op: cfData?.data?.[key] ? 'replace' : 'add',
+          path: `/data/${key}`,
+          value: settings,
+        },
+      ];
+      if (settings !== cfData.data?.[key]) {
+        (async () => {
+          await patchConfigMap(patch);
+        })();
+      }
+    }
+    // This effect should only be run on change of settings state.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [settings]);
+
+  return [settings, setSettings] as const;
```

Comment:
Since settings can be lazy loaded, I think we need another return value which indicates whether the settings have been fetched yet or not.

Reply:
Valid point, esp when we mean defaultValue is returned != data fetched